modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C

```diff
@@ -0,0 +1,101 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdaptiveImportanceStats.h"
+#include "Sampler.h"
+#include "Distribution.h"
+#include "Normal.h"
+#include "libmesh/utility.h"
+
+registerMooseObject("StochasticToolsApp", AdaptiveImportanceStats);
+
+InputParameters
+AdaptiveImportanceStats::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Reporter to compute statistics corresponding to the "
+                             "AdaptiveImportanceSampler.");
+  params.addRequiredParam<ReporterName>("output_value",
+                                        "Value of the model output from the SubApp.");
+  params.addParam<ReporterValueName>("mu_imp", "mu_imp", "Means of the importance distributions.");
+  params.addParam<ReporterValueName>(
+      "std_imp", "std_imp", "Standard deviations of the importance distributions.");
+  params.addParam<ReporterValueName>("pf", "pf", "Failure probability estimate.");
+  params.addParam<ReporterValueName>(
+      "cov_pf", "cov_pf", "Coefficient of variation of failure probability.");
+  params.addRequiredParam<SamplerName>("sampler", "The sampler object.");
+  return params;
+}
+
+AdaptiveImportanceStats::AdaptiveImportanceStats(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _output_value(getReporterValue<std::vector<Real>>("output_value", REPORTER_MODE_DISTRIBUTED)),
+    _mu_imp(declareValue<std::vector<Real>>("mu_imp")),
+    _std_imp(declareValue<std::vector<Real>>("std_imp")),
+    _pf(declareValue<std::vector<Real>>("pf")),
+    _cov_pf(declareValue<std::vector<Real>>("cov_pf")),
+    _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
+    _sampler(getSampler("sampler")),
+    _ais(dynamic_cast<const AdaptiveImportanceSampler *>(&_sampler)),
+    _check_step(std::numeric_limits<int>::max())
+{
+  if (!_ais)
+    paramError("sampler", "The selected sampler is not of type AdaptiveImportance.");
+
+  // Initialize variables
+  const auto rows = _sampler.getNumberOfRows();
+  _mu_imp.resize(rows);
+  _std_imp.resize(rows);
+  _pf.resize(1);
+  _cov_pf.resize(1);
+  _pf_sum = 0.0;
+  _var_sum = 0.0;
+  _distributions_store = _ais->getDistributionNames();
+  _factor = _ais->getStdFactor();
+}
+
+void
+AdaptiveImportanceStats::execute()
+{
+  if (_sampler.getNumberOfLocalRows() == 0 || _check_step == _step)
```

Comment:
We should probably figure out what will happen in parallel when a processor doesn't own a row.